Rules/AvoidOverwritingBuiltInCmdlets.cs

```diff
@@ -0,0 +1,284 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Management.Automation.Language;
+using System.Text.RegularExpressions;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// AvoidOverwritingBuiltInCmdlets: Checks if a script overwrites a cmdlet that comes with PowerShell
+    /// </summary>
+#if !CORECLR
+    [Export(typeof(IScriptRule))]
+#endif
+    /// <summary>
+    /// A class to check if a script overwrites a cmdlet that comes with PowerShell
+    /// </summary>
+    public class AvoidOverwritingBuiltInCmdlets : ConfigurableRule
+    {
+        [ConfigurableRuleProperty(defaultValue: "")]
+        public string[] PowerShellVersion { get; set; }
+        private Dictionary<string, HashSet<string>> cmdletMap;
+        private bool initialized;
+
+
+        /// <summary>
+        /// Construct an object of AvoidOverwritingBuiltInCmdlets type.
+        /// </summary>
+        public AvoidOverwritingBuiltInCmdlets() : base()
+        {
+            initialized = false;
+            cmdletMap = new Dictionary<string, HashSet<string>>();
+            Enable = true;  // Enable rule by default
+            
+            string versionTest = string.Join("", PowerShellVersion);
+
+            if (versionTest != "core-6.1.0-windows" && versionTest != "desktop-5.1.14393.206-windows")
+            {
+                // PowerShellVersion is not already set to one of the acceptable defaults
+                // Try launching `pwsh -v` to see if PowerShell 6+ is installed, and use those cmdlets
+                // as a default. If 6+ is not installed this will throw an error, which when caught will
+                // allow us to use the PowerShell 5 cmdlets as a default.
+                try
```

Comment:
I've moved this out of the constructor. I'm not 100% sure if this is the best place for it now, since there's a lot about how the PSSA engine works that I don't know, so it might need to move again....